src/main/kotlin/org/elm/workspace/compiler/ElmBuildAction.kt

```diff
@@ -0,0 +1,95 @@
+package org.elm.workspace.compiler
+
+import com.intellij.execution.ExecutionException
+import com.intellij.notification.NotificationType
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.MessageType
+import com.intellij.openapi.ui.popup.Balloon
+import com.intellij.openapi.ui.popup.JBPopupFactory
+import com.intellij.openapi.wm.ToolWindowManager
+import com.intellij.openapi.wm.WindowManager
+import com.intellij.ui.awt.RelativePoint
+import com.intellij.util.messages.Topic
+import org.elm.ide.notifications.showBalloon
+import org.elm.lang.core.lookup.ClientLocation
+import org.elm.lang.core.lookup.ElmLookup
+import org.elm.lang.core.psi.ElmNamedElement
+import org.elm.lang.core.types.TyUnion
+import org.elm.lang.core.types.findInference
+import org.elm.openapiext.saveAllDocuments
+import org.elm.workspace.ElmProject
+import org.elm.workspace.elmToolchain
+import org.elm.workspace.elmWorkspace
+
+
+class ElmBuildAction : AnAction() {
+
+    private val elmMainTypes: Set<Pair<String, String>> = hashSetOf(Pair("Platform", "Program"), Pair("Html", "Html"))
+
+    private val elmJsonReport = ElmJsonReport()
+
+    override fun actionPerformed(e: AnActionEvent) {
+        val project = e.project
+                ?: return
+        saveAllDocuments()
+
+        val elmCLI = project.elmToolchain.elmCLI
+        val fixAction = "Fix" to { project.elmWorkspace.showConfigureToolchainUI() }
+        if (elmCLI == null) {
+            project.showBalloon("Could not find elm", NotificationType.ERROR, fixAction)
+            return
+        }
+
+        val elmProject = project.elmWorkspace.allProjects.first()
+
+        // find "main" function
+        val clientLocation = LookupClientLocation(project, elmProject)
+        val elements = ElmLookup.findByName<ElmNamedElement>("main", clientLocation)
+        val mainElements = elements
+                .filter { element -> element.findInference() != null && element.findInference()?.ty is TyUnion }
+                .filter { element -> elmMainTypes.contains(Pair((element.findInference()?.ty as TyUnion).module, (element.findInference()?.ty as TyUnion).name)) }
+
+        if (mainElements.isNotEmpty()) {
+            val path = mainElements[0].containingFile.virtualFile.path
+            try {
```

Comment:
The `try/catch` should just cover the call to `elmCLI.make()` as that is the only part that can throw an `ExecutionException`. Also, Kotlin's try/catch can be used as an expression (e.g. on the right-hand side of the assignment to the local `json` variable), which may be nicer here.